### PR TITLE
NPFL128 Project - Torlak accent prediction

### PR DESCRIPTION
**Pull request by toni5rovic**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,5 +2,147 @@
 
 ## NPFL128 - Language Technologies in Practice Project
 
-Author: Antonije Petrović, Charles University in Prague
+Antonije Petrović, Charles University in Prague
```
Comment (jirka-x1) [inline README.md:6]: Add one sentence explaining what the project does (something like the name of this pull request). I know you explain it below, but it should be clear before I need to read introduction.

File: README.md [hunk 2]
```diff
@@ -2,5 +2,147 @@
...
+| 0007 | 15     | 256        | 0.5     | 75            | 16     | Yes   | 256, 128, 64      | 0.001                   | 95.04%       |
+| 0008 | 15     | 256        | 0.5     | 64            | 12     | Yes   | 512, 256, 128, 64 | 0.001                   | 95.24%       |
+| 0009 | 20     | 64         | 0.5     | 62            | 6      | Yes   | 256, 128, 16      | 0.001 -> 0.0001         | 95.70%       |
+| 0010 | 20     | 128        | 0.5     | 62            | 6      | Yes   | 512, 128, 16      | 0.001 -> 0.0001         | 95.80%       |
+| 0011 | 20     | 64         | 0.5     | 62            | 8      | Yes   | 128, 64, 16       | 0.001 -> 0.0001         | 95.47%       |
+| 0012 | 20     | 512        | 0.4     | 80            | 10     | Yes   | 2048, 512, 16     | 0.001 -> 0.0001         | 95.25%       |
+| 0013 | 20     | 512        | 0.4     | 80            | 10     | Yes   | 2048, 1028, 32    | 0.001 -> 0.0001         | 95.22%       |
+
+We can see that increasing the width of the network helped and we get accuracy up to 95.80% on the validation dataset.
+This experiment (ID 10) gives an accuracy of 98.10% on the test dataset.
```
Comment (jirka-x1) [inline README.md:97]: Do I understand correctly that the test dataset gives you 2.3%-pts better results than the validation dataset?

File: torlakian_data.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                len(predictions), len(gold)))
+
+        correct = 0
+        for i in range(len(gold)):
+            if predictions[i].lower() != gold[i].lower():
+                raise RuntimeError("The predictions and gold data differ on position {}: {} vs {}.".format(
+                    i, repr(predictions[i:i + 20].lower()), repr(gold[i:i + 20].lower())))
+
+            correct += gold[i] == predictions[i]
+        return 100 * correct / len(gold)
```
Comment (jirka-x1) [inline torlakian_data.py]: Are you sure you are gettin the right results? This looks like interger division.  I would write:
`100.0 * correct / len(gold)`

File: torlakian_data.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    test: Dataset
+
+    # Evaluation infrastructure.
+    @staticmethod
+    def evaluate(gold_dataset: Dataset, predictions: str) -> float:
+        gold = gold_dataset.text
+
+        if len(predictions) < len(gold):
+            raise RuntimeError("The predictions are shorter than gold data: {} vs {}.".format(
+                len(predictions), len(gold)))
```
Comment (jirka-x1) [inline torlakian_data.py]: it is better to use f-strings:
```suggestion
            raise RuntimeError(f"The predictions are shorter than gold data: {len(predictions)} vs {len(gold)}.")
```

File: torlakian_data.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    URL = 'https://www.clarin.si/repository/xmlui/handle/11356/1281/allzip'
+
+
+    # Class for working with the data in a customized way. 
+    # - window: number of characters left and right from the current position
+    # - alphabet: maximum number of most frequent characters that are going to stay in the data,
+    #   while the rest are replaced by <unk>
+    # - clean: boolean that states if the data should be cleaned of annotator's marks
+    # - seed: for setting fixed random seed
+    class Dataset:
```
Comment (jirka-x1) [inline torlakian_data.py:35]: Add type hints to arguments and return values of each method. It will help to understand them.

File: torlakian_data.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+                labels[i] = self._text[i].isupper()
+            self._data = {"windows": windows, "labels": labels}
+
+            # Compute alphabet
+            self._alphabet = [None] * len(alphabet_map)
+            for key, value in alphabet_map.items():
+                self._alphabet[value] = key
+
+        # Cleans the data of annotator notes in the speaker utterances
+        def clean_data(self, data):
```
Comment (jirka-x1) [inline torlakian_data.py]:
```suggestion
        def clean_data(self, data: str ) -> str:
```

File: torlak_accent.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+import numpy as np
+import tensorflow as tf
+import time
+from torlakian_data import TorlakianData
+
+os.environ['CUDA_VISIBLE_DEVICES'] = '-1'
+
+# Based on: https://github.com/ufal/npfl114/blob/master/labs/03/uppercase.py
+
+parser = argparse.ArgumentParser()
```
Comment (jirka-x1) [inline torlak_accent.py:16]: be consistent. In the file below, you have argparser as part of __main__.

File: model.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        predictions = map(do_uppercase, range(len(output)))
+        output_text = ''.join(predictions)
+        predictions_file.write(output_text)
+
+        end_time = time.time()
+        print(f"Prediction took: {end_time - start_time} s")
+
+if __name__ == "__main__":
+    args = parser.parse_args([] if "__file__" not in globals() else None)
+    main(args)
```
Comment (jirka-x1) [inline model.py]: See III.7 https://ufal.mff.cuni.cz/courses/npfl128/project

File: torlak_accent.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        tf.keras.layers.Dense(32, activation=tf.nn.relu),
+        tf.keras.layers.Dropout(DROPOUT),
+        tf.keras.layers.Dense(1, activation=tf.nn.sigmoid)
+    ]
+
+    model = tf.keras.Sequential(layers)
+
+    model.load_weights(args.model)
+    
+    def do_uppercase(index):
```
Comment (jirka-x1) [inline torlak_accent.py]: you have a very similar function above. what is the reason to have them both?

File: torlak_accent.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            lines = input_file.readlines()
+            input_text = ''.join(lines)
+    else:
+        input_text = args.text
+
+    dataset = torlakian_data.Dataset(input_text, WINDOW, ALPHABET_SIZE, CLEAN)
+    output = model(dataset.data["windows"])
+    output = output > 0.5
+
+    predictions = map(do_uppercase, range(len(output)))
```
Comment (jirka-x1) [inline torlak_accent.py]: this is confusing, do_uppercase runs on indexes of `dataset.text` but map uses indexes of `output`. They might be the same size, but it is not obvious.

File: model.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    logs = model.fit(
+        torlakian_data.train.data["windows"], 
+        torlakian_data.train.data["labels"],
+        batch_size=args.batch_size, 
+        epochs=args.epochs,
+        validation_data=(torlakian_data.val.data["windows"], torlakian_data.val.data["labels"]),
+        callbacks=[tb_callback, stop_early],
+    )
+
+    def do_uppercase(index):
```
Comment (jirka-x1) [inline model.py]: add comment on thsi (why `...test.alphabet`)

File: torlakian_data.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+
+        with open('data/train.txt', 'w') as f:
+            f.write('\n'.join(train))
+        with open('data/val.txt', 'w') as f:
+            f.write('\n'.join(val))
+        with open('data/test.txt', 'w') as f:
+            f.write('\n'.join(test))
+
+    def __init__(self, window: int, alphabet_size: int = 0, clean: bool = False):
+        path = self.download()
```
Comment (jirka-x1) [inline torlakian_data.py]: shouldn't these three funcstions be private? Or do you also expect them to be called outside of `__init__` ?

File: torlakian_data.py [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+        self.unzip(path)
+        self.join_data()
+
+        for dataset in ["train", "val", "test"]:
```
Comment (jirka-x1) [inline torlakian_data.py:218]: `val` or `dev` ?

File: README.md (hunk 2 above)
Comment (toni5rovic) [reply README.md:97]: Yes! I guess it is possible to happen because test set is "easier" than the validation dataset.

File: torlakian_data.py (hunk 3 above)
Comment (toni5rovic) [reply torlakian_data.py]: Yes, the results are correct. It seems that Python is smart enough but I agree it should be `100.0` for more clarity.

File: torlakian_data.py (hunk 4 above)
Comment (toni5rovic) [reply torlakian_data.py]: fixed

File: README.md (hunk 1 above)
Comment (toni5rovic) [reply README.md:6]: fixed

File: torlakian_data.py (hunk 5 above)
Comment (toni5rovic) [reply torlakian_data.py:35]: fixed

File: torlakian_data.py (hunk 12 above)
Comment (toni5rovic) [reply torlakian_data.py]: Agreed, they should be private. Fixed

File: torlakian_data.py (hunk 13 above)
Comment (toni5rovic) [reply torlakian_data.py:218]: Changed to `val` everywhere

File: model.py (hunk 11 above)
Comment (toni5rovic) [reply model.py]: It was quite confusing, I agree. I parameterized the function and moved to the file `inference.py`. Hopefully it's more readable now.

File: torlak_accent.py (hunk 10 above)
Comment (toni5rovic) [reply torlak_accent.py]: Understood. I added comments in the method to clarify this.

File: torlak_accent.py (hunk 9 above)
Comment (toni5rovic) [reply torlak_accent.py]: Fixed. Moved the function to file `inference.py` and now it makes more sense.

File: model.py (hunk 8 above)
Comment (toni5rovic) [reply model.py]: Fixed

File: torlak_accent.py (hunk 7 above)
Comment (toni5rovic) [reply torlak_accent.py:16]: Not sure which file you are referring to. In the file `model.py` the situation is the same: ArgumentParse is instantiated outside of main, and used as an argument to the `main` method. Right now, `model.py` and `torlak_accent.py` are the same in that regard.

Comment (jirka-x1) [reply torlak_accent.py:16]: Yes but look at `torlakian_data.py` There it is under __main__

Comment (toni5rovic) [reply torlak_accent.py:16]: Fixed
